salt/modules/zypperpkg.py

```diff
@@ -218,6 +222,21 @@ def stderr(self):
     def pid(self):
         return self.__call_result.get('pid', '')
 
+    def refresh_zypper_flags(self):
+        try:
+            zypp_version = version('zypper')
+            # zypper version 1.11.34 in SLE12 update supports vendor change for only dist upgrade
+            if version_cmp(zypp_version, '1.11.34') >= 0:
+                # zypper version supports vendor change for dist upgrade
+                self.dup_avc = True
+            # zypper version 1.14.8 in SLE15 update supports vendor change in install/patch/upgrading
+            if version_cmp(zypp_version, '1.14.8') >= 0:
+                self.inst_avc = True
```

Comment:
```python
self.dup_avc = version_cmp(zypp_version, "1.11.34") > -1
self.inst_avc = version_cmp(zypp_version, "1.14.8") > -1
```

Still needs refactoring on better version-to-feature tracking.

Reply:
@isbm I'm not sure if I'm getting you proposal here. Why is this the suggested change improving the "version-to-feature tracking"? Or is this unrelated?

Reply:
Change is only makes less codelines.

The problem I am referring to, is that you still have scattered hard-coded versions (and these will grow up), mapped to... nothing. So you will end up with "version x.y.z does this, that and that", "version  x.y.y does only this, but not that" etc etc. You don't want that. @meaksh  had a good idea with capabilities and I wrote in 2018 an RFC and implementation of remote inspection so you can construct Jinja the way it "knows" if a particular feature is there or not. Because bare Salt versions are not that helpful if you patch faster then rolling the releases (which is usually the case).

There is another problem you should solve: "Can client alice.acme.com perform feature X and bob.acme.com not?" — and before you even go there. Otherwise you will often end up with a semi-failed-but-still-good states, which you will still need to accept: something that e.g. Ansible does not have, and it is very good.

So for now you could do as-is: hardcoding versions. But in a long run the problem is still not solved.

@meaksh your time. :)

Reply:
Hey @isbm :)
I see we're pointing two different problems here:

1) Make `zypperpkg` to properly support `allow-vendor-change` flag, which are missing or broken atm (which is the purpose of this PR)
2) Determine the capabilties of a certain minion in the Jinja (SLS) context .

AFAIU this PR is tackling problem `1` but not problem `2`.

I see your point that we don't really want to continue growing hardcoded Zypper versions checks on every new Zypper feature we want to implement in Salt. This would need a better thinking and refactor.

About problem `2`, this is something to keep thinking, but I think we shouldn't make this PR to require it. I forgot about https://github.com/saltstack/salt/pull/48152 :sweat_smile: 

Since we have been dealing with this PR since a while already, and it's currently blocking other implementations. I would say that we could accept this current implementation now, and continue with the refactor in a following-up PR.

Reply:
@meaksh yes, exactly. The problem `#1` is just "remove these if/else, but add two lines and you're done for now". But *also* I pointed out so for the *future* we need to keep in mind the problem `#2`, as you correctly pointed out.